File: bridges/python/src/sdk/tools/bash/bash_tool.py
import os
from pathlib import Path
from typing import Dict, Any, Optional
from ...base_tool import BaseTool, ExecuteCommandOptions
from ...toolkit_config import ToolkitConfig

DEFAULT_SETTINGS = {}
REQUIRED_SETTINGS = []

CRITICAL_COMMAND_PATTERNS = [
    "rm -rf /",
    "rm -rf /*",
    "mkfs",
    "format",
    "fdisk",
    "kill -9 -1",
]

HIGH_RISK_COMMAND_PATTERNS = [
    "dd if=",
    "eval $(curl",
    "eval $(wget",
]

MEDIUM_RISK_COMMAND_PATTERNS = []

UNSAFE_COMMAND_PATTERNS = [
    *CRITICAL_COMMAND_PATTERNS,
    *HIGH_RISK_COMMAND_PATTERNS,
    "fork()",
    "while true; do",
]


class BashTool(BaseTool):
    TOOLKIT = "operating_system_control"

    def __init__(self):
        super().__init__()
        self.config = ToolkitConfig.load(self.TOOLKIT, self.tool_name)
        self.settings = ToolkitConfig.load_tool_settings(
            self.TOOLKIT, self.tool_name, DEFAULT_SETTINGS
        )
        self.required_settings = REQUIRED_SETTINGS
        self._check_required_settings(self.tool_name)

    @property
    def tool_name(self) -> str:
        return "bash"

    @property
    def toolkit(self) -> str:
        return self.TOOLKIT

    @property
    def description(self) -> str:
        return self.config["description"]

    def execute_bash_command(
        self,
        command: str,
        cwd: Optional[str] = None,
        timeout: Optional[int] = 30,
        capture_output: bool = True,
    ) -> Dict[str, Any]:
        """
        Execute a bash command and return the result.

        Args:
            command: The bash command to execute
            cwd: Working directory for the command
            timeout: Command timeout in seconds (default: 30)
            capture_output: Whether to capture stdout/stderr (default: True)

        Returns:
            Dict with keys: success, stdout, stderr, returncode, command
        """
        analyzed_command = self._resolve_command_for_safety_analysis(command)
        if not self.is_safe_command(analyzed_command):
            risk_level = self.get_command_risk_level(analyzed_command)
            risk_description = self.get_risk_description(analyzed_command)
            return {
                "success": False,
                "stdout": "",
                "stderr": f"Blocked unsafe bash command ({risk_level} risk): This command may {risk_description}.",
                "returncode": -1,
                "command": command,
            }

        try:
            # Use shell execution via the base tool's command execution method
            # Split command into binary and args for proper execution
            # For bash commands, we'll use 'bash' as the binary and '-c' with the command as args
            result_output = self.execute_command(
                ExecuteCommandOptions(
                    binary_name="bash",
                    args=["-c", command],
                    options={
                        "sync": True,
                        "cwd": cwd or os.getcwd(),
                        "timeout": timeout,
                    },
                    skip_binary_download=True,  # bash is a built-in command, no need to download
                )
            )

            return {
                "success": True,
                "stdout": result_output.strip(),
                "stderr": "",
                "returncode": 0,
                "command": command,
            }

        except Exception as e:
            error_message = str(e)

            # Parse error to determine if it was a timeout, command failure, or other error
            if "timed out" in error_message.lower():
                return {
                    "success": False,
                    "stdout": "",
                    "stderr": f"Command timed out after {timeout} seconds",
                    "returncode": -1,
                    "command": command,
                }
            elif "failed with exit code" in error_message:
                # Extract exit code and error from the base tool's error message
                import re

                exit_code_match = re.search(r"exit code (\d+)", error_message)
                exit_code = int(exit_code_match.group(1)) if exit_code_match else -1

                # Extract stderr from the error message if present
                stderr_match = re.search(r"exit code \d+: (.+)$", error_message)
                stderr = stderr_match.group(1) if stderr_match else error_message

                return {
                    "success": False,
                    "stdout": "",
                    "stderr": stderr,
                    "returncode": exit_code,
                    "command": command,
                }
            else:
                return {
                    "success": False,
                    "stdout": "",
                    "stderr": error_message,
                    "returncode": -1,
                    "command": command,
                }

    def is_safe_command(self, command: str) -> bool:
        """
        Basic safety check for bash commands.
        Returns True if command appears safe to execute.
        """
        command_lower = command.lower()

        for pattern in UNSAFE_COMMAND_PATTERNS:
            if pattern in command_lower:
                return False

        if self._is_download_piped_to_shell(command_lower):
            return False

        return True

    def get_command_risk_level(self, command: str) -> str:
        """
        Assess the risk level of a command.
        Returns: 'low', 'medium', 'high', or 'critical'
        """

        command_lower = command.lower()

        risk_level = "low"
        for pattern in CRITICAL_COMMAND_PATTERNS:
            if pattern in command_lower:
                risk_level = "critical"
                break

        if risk_level == "low":
            for pattern in HIGH_RISK_COMMAND_PATTERNS:
                if pattern in command_lower:
                    risk_level = "high"
                    break

        if risk_level == "low" and self._is_download_piped_to_shell(command_lower):
            risk_level = "high"

        if risk_level == "low":
            for pattern in MEDIUM_RISK_COMMAND_PATTERNS:
                if pattern in command_lower:
                    risk_level = "medium"
                    break

        return risk_level

    def get_risk_description(self, command: str) -> str:
        """
        Get a human-readable description of the command's risk.
        """
        risk_level = self.get_command_risk_level(command)
        command_lower = command.lower()

        if "rm" in command_lower:
            return "delete files or directories permanently"
        elif "sudo" in command_lower:
            return "make system-level changes with elevated privileges"
        elif "kill" in command_lower:
            return "terminate running processes"
        elif "chmod" in command_lower or "chown" in command_lower:
            return "change file permissions or ownership"
        elif any(pkg in command_lower for pkg in ["apt", "yum", "brew", "pip"]):
            return "install or modify system packages"
        elif self._is_download_piped_to_shell(command_lower):
            return "download remote content and execute it as a shell script"
        elif "curl" in command_lower or "wget" in command_lower:
            return "download content from the internet"
        else:
            return {
                "critical": "cause severe system damage",
                "high": "cause significant system changes",
                "medium": "modify your system",
                "low": "perform system operations",
            }.get(risk_level, "affect your system")

    def _resolve_command_for_safety_analysis(self, command: str) -> str:
        trimmed_command = command.strip()
        if not trimmed_command or any(char.isspace() for char in trimmed_command):
            return command

        candidate_path = Path(trimmed_command).expanduser().resolve()

        try:
            if not candidate_path.is_file():
                return command

            file_content = candidate_path.read_text(encoding="utf-8")
            return file_content if file_content.strip() else command
        except Exception:
            return command

    def _is_download_piped_to_shell(self, command_lower: str) -> bool:
        downloads_remote_content = "curl" in command_lower or "wget" in command_lower
        pipes_to_shell = "| bash" in command_lower or "| sh" in command_lower
        return downloads_remote_content and pipes_to_shell

File: bridges/nodejs/src/sdk/tools/bash/bash-tool.ts
import fs from 'node:fs'
import path from 'node:path'

import { Tool } from '@sdk/base-tool'
import { ToolkitConfig } from '@sdk/toolkit-config'

const DEFAULT_SETTINGS: Record<string, unknown> = {}
const REQUIRED_SETTINGS: string[] = []

interface BashResult {
  success: boolean
  stdout: string
  stderr: string
  returncode: number
  command: string
}

interface ExecuteOptions {
  cwd?: string
  timeout?: number
  captureOutput?: boolean
}

const CRITICAL_COMMAND_PATTERNS = [
  'rm -rf /',
  'rm -rf /*',
  'mkfs',
  'format',
  'fdisk',
  'kill -9 -1'
]

const HIGH_RISK_COMMAND_PATTERNS = [
  'dd if=',
  'eval $(curl',
  'eval $(wget'
]

const MEDIUM_RISK_COMMAND_PATTERNS: string[] = []

const UNSAFE_COMMAND_PATTERNS = [
  ...CRITICAL_COMMAND_PATTERNS,
  ...HIGH_RISK_COMMAND_PATTERNS,
  'fork()',
  'while true; do'
]

export default class BashTool extends Tool {
  private static readonly TOOLKIT = 'operating_system_control'
  private readonly config: ReturnType<typeof ToolkitConfig.load>

  constructor() {
    super()
    // Load configuration from central toolkits directory
    this.config = ToolkitConfig.load(BashTool.TOOLKIT, this.toolName)
    const toolSettings = ToolkitConfig.loadToolSettings(
      BashTool.TOOLKIT,
      this.toolName,
      DEFAULT_SETTINGS
    )
    this.settings = toolSettings
    this.requiredSettings = REQUIRED_SETTINGS
    this.checkRequiredSettings(this.toolName)
  }

  get toolName(): string {
    return 'bash'
  }

  get toolkit(): string {
    return BashTool.TOOLKIT
  }

  get description(): string {
    return this.config['description']
  }

  /**
   * Execute a bash command and return the result.
   */
  async executeBashCommand(
    command: string,
    options: ExecuteOptions = {}
  ): Promise<BashResult> {
    const { cwd = process.cwd(), timeout = 30 } = options
    const analyzedCommand = await this.resolveCommandForSafetyAnalysis(command)
    const isSafe = await this.isSafeCommand(analyzedCommand)

    if (!isSafe) {
      const riskLevel = await this.getCommandRiskLevel(analyzedCommand)
      const riskDescription = await this.getRiskDescription(analyzedCommand)

      return {
        success: false,
        stdout: '',
        stderr: `Blocked unsafe bash command (${riskLevel} risk): This command may ${riskDescription}.`,
        returncode: -1,
        command
      }
    }

    try {
      // Use the base tool's command execution method
      // For bash commands, we'll use 'bash' as the binary and '-c' with the command as args
      const resultOutput = await this.executeCommand({
        binaryName: 'bash',
        args: ['-c', command],
        options: {
          sync: true,
          cwd,
          timeout: timeout * 1_000
        },
        skipBinaryDownload: true // bash is a built-in command, no need to download
      })

      return {
        success: true,
        stdout: resultOutput.trim(),
        stderr: '',
        returncode: 0,
        command
      }
    } catch (error: unknown) {
      const errorMessage = (error as Error).message

      // Parse error to determine if it was a timeout, command failure, or other error
      if (errorMessage.toLowerCase().includes('timed out')) {
        return {
          success: false,
          stdout: '',
          stderr: `Command timed out after ${timeout} seconds`,
          returncode: -1,
          command
        }
      } else if (errorMessage.includes('failed with exit code')) {
        // Extract exit code and error from the base tool's error message
        const exitCodeMatch = errorMessage.match(/exit code (\d+)/)
        const exitCode =
          exitCodeMatch && exitCodeMatch[1]
            ? parseInt(exitCodeMatch[1], 10)
            : -1

        // Extract stderr from the error message if present
        const stderrMatch = errorMessage.match(/exit code \d+: (.+)$/)
        const stderr =
          stderrMatch && stderrMatch[1] ? stderrMatch[1] : errorMessage

        return {
          success: false,
          stdout: '',
          stderr,
          returncode: exitCode,
          command
        }
      } else {
        return {
          success: false,
          stdout: '',
          stderr: errorMessage,
          returncode: -1,
          command
        }
      }
    }
  }

  /**
   * Basic safety check for bash commands.
   * Returns True if command appears safe to execute.
   */
  async isSafeCommand(command: string): Promise<boolean> {
    const commandLower = command.toLowerCase()

    for (const pattern of UNSAFE_COMMAND_PATTERNS) {
      if (commandLower.includes(pattern)) {
        return false
      }
    }

    if (this.isDownloadPipedToShell(commandLower)) {
      return false
    }

    return true
  }

  /**
   * Assess the risk level of a command.
   * Returns: 'low', 'medium', 'high', or 'critical'
   */
  async getCommandRiskLevel(command: string): Promise<string> {
    const commandLower = command.toLowerCase()

    let riskLevel = 'low'

    for (const pattern of CRITICAL_COMMAND_PATTERNS) {
      if (commandLower.includes(pattern)) {
        riskLevel = 'critical'
        break
      }
    }

    if (riskLevel === 'low') {
      for (const pattern of HIGH_RISK_COMMAND_PATTERNS) {
        if (commandLower.includes(pattern)) {
          riskLevel = 'high'
          break
        }
      }
    }

    if (riskLevel === 'low' && this.isDownloadPipedToShell(commandLower)) {
      riskLevel = 'high'
    }

    if (riskLevel === 'low') {
      for (const pattern of MEDIUM_RISK_COMMAND_PATTERNS) {
        if (commandLower.includes(pattern)) {
          riskLevel = 'medium'
          break
        }
      }
    }

    return riskLevel
  }

  /**
   * Get a human-readable description of the command's risk.
   */
  async getRiskDescription(command: string): Promise<string> {
    const riskLevel = await this.getCommandRiskLevel(command)
    const commandLower = command.toLowerCase()

    if (commandLower.includes('rm')) {
      return 'delete files or directories permanently'
    } else if (commandLower.includes('sudo')) {
      return 'make system-level changes with elevated privileges'
    } else if (commandLower.includes('kill')) {
      return 'terminate running processes'
    } else if (
      commandLower.includes('chmod') ||
      commandLower.includes('chown')
    ) {
      return 'change file permissions or ownership'
    } else if (
      ['apt', 'yum', 'brew', 'pip'].some((pkg) => commandLower.includes(pkg))
    ) {
      return 'install or modify system packages'
    } else if (this.isDownloadPipedToShell(commandLower)) {
      return 'download remote content and execute it as a shell script'
    } else if (commandLower.includes('curl') || commandLower.includes('wget')) {
      return 'download content from the internet'
    } else {
      const descriptions: Record<string, string> = {
        critical: 'cause severe system damage',
        high: 'cause significant system changes',
        medium: 'modify your system',
        low: 'perform system operations'
      }
      return descriptions[riskLevel] || 'affect your system'
    }
  }

  private async resolveCommandForSafetyAnalysis(command: string): Promise<string> {
    const trimmedCommand = command.trim()
    if (!trimmedCommand || /\s/.test(trimmedCommand)) {
      return command
    }

    const resolvedPath = path.resolve(trimmedCommand)

    try {
      const stats = await fs.promises.stat(resolvedPath)
      if (!stats.isFile()) {
        return command
      }

      const fileContent = await fs.promises.readFile(resolvedPath, 'utf8')
      if (!fileContent.trim()) {
        return command
      }

      return fileContent
    } catch {
      return command
    }
  }

  private isDownloadPipedToShell(commandLower: string): boolean {
    const downloadsRemoteContent =
      commandLower.includes('curl') || commandLower.includes('wget')
    const pipesToShell =
      commandLower.includes('| bash') || commandLower.includes('| sh')

    return downloadsRemoteContent && pipesToShell
  }
}
